### PR TITLE
Feature/upload

### PR DESCRIPTION
**Pull request by egrace479**
Added upload option to Dashboard (csv or xls).
- Updated functions: 
  - `get_filenames`: returns filenames and paths, also patches "Any" subspecies glitch (if user doesn't remove `Any` option before selecting specific subspecies and pressing "Display Images", it will just return images of the selected subspecies).
  -  `test_get_filenames` and `get_images` updated to interact with updated `get_filenames`.
  - `get_data`: also takes boolean indicating if latitude & longitude values are in data to do `lat-lon` manipulations and sets `locality` column with `lat-lon` values if it's not already a column in the dataframe. 
  - `test_get_data`: now checks two variations, one with lat & long, one without.
- Also removed additional taxa info from cat_list (in our test data it was just subspecies data repeated).
- Updated README to reflect upload requirements and added two datasets for testing/demo.
Resolves Issue #23

### CODE REVIEW

**Comment by egrace479** [comment]
This is currently what it looks like before upload, then below the line gets replaced by the dashboard as it has been. Clicking on "Upload Data" again allows for a new upload.

Main Page

Should I add the column requirements underneath as well? It currently will throw an error (missing column, wrong file type, etc.) and point the user to [how-it-works](https://github.com/Imageomics/dashboard-prototype#how-it-works):
Error Message (missing column)

File: README.md [hunk 1]
```diff
@@ -1,9 +1,24 @@
 # Dashboard Prototype
-Prototype data dashboard using the [Cuthill Gold Standard Dataset](https://datacommons.tdai.osu.edu/dataset.xhtml?persistentId=doi:10.5072/FK2/GZYWNV&version=DRAFT), which was processed from Cuthill, et. al. (original dataset available at [doi:10.5061/dryad.2hp1978](https://doi.org/10.5061/dryad.2hp1978)).
+Prototype data dashboard using the [Cuthill Gold Standard Dataset](https://huggingface.co/datasets/imageomics/Curated_GoldStandard_Hoyal_Cuthill), which was processed from Cuthill, et. al. (original dataset available at [doi:10.5061/dryad.2hp1978](https://doi.org/10.5061/dryad.2hp1978)). Test datasets (the processed version of Cuthill's data with and without filepath URLs) are available in [test_data](https://github.com/Imageomics/dashboard-prototype/tree/main/test_data).
```
Comment (hlapp) [inline README.md:2]:
```suggestion
Prototype data dashboard using the [Cuthill Gold Standard Dataset](https://huggingface.co/datasets/imageomics/Curated_GoldStandard_Hoyal_Cuthill), which was processed from Cuthill, et. al. (original dataset available at [doi:10.5061/dryad.2hp1978](https://doi.org/10.5061/dryad.2hp1978)). Test datasets (the processed version of Cuthill's data with and without filepath URLs) are available in [test_data](./test_data).
```

It's probably better not to tie this to a particular branch, and instead just use the relative path.

**Comment by hlapp** [comment]
> This is currently what it looks like before upload, then below the line gets replaced by the dashboard as it has been. [...]
> Should I add the column requirements underneath as well? It currently will throw an error (missing column, wrong file type, etc.) and point the user to [how-it-works](https://github.com/Imageomics/dashboard-prototype#how-it-works)

It seems helpful to provide a link to "How it works" before someone runs into an error unintentionally because they didn't fully understand which columns they'd have to have.

**Review by hlapp — COMMENTED** [review]
It seems this change set lumps together changes for multiple separate and not necessarily interrelated issues. Generally I'd suggest breaking changes up into meaningful and coherent pieces, this makes review and feedback much easier.

Overall this looks fine but I notice there's a fair amount of code duplication. In the conditional for whether geo-coordinates are available or not, it seems only a small part differs for the two cases, and the rest is duplicated. There are also strings or string prefixes (such as `https://github.com/Imageomics/dashboard-prototype/raw/main/test_data/images/`) that occur many times over, and could just be defined at the beginning and then reused.

**Comment by egrace479** [comment]
> It seems this change set lumps together changes for multiple separate and not necessarily interrelated issues. Generally I'd suggest breaking changes up into meaningful and coherent pieces, this makes review and feedback much easier.
> 
> Overall this looks fine but I notice there's a fair amount of code duplication. In the conditional for whether geo-coordinates are available or not, it seems only a small part differs for the two cases, and the rest is duplicated. There are also strings or string prefixes (such as `https://github.com/Imageomics/dashboard-prototype/raw/main/test_data/images/`) that occur many times over, and could just be defined at the beginning and then reused.

For this PR, would it have been better to do the file retrieval generalization with a revised preloaded dataset and test as its own PR, then move the main div out separately and add the upload feature?

To your second point, I can adjust the geo-coordinate if statement to append the final div to `hist_div` dependent on the boolean. For the string duplication, are you referring to `test_query`? I can adjust that as well.

**Comment by egrace479** [comment]
Updated as per @hlapp's suggestions.

New landing page:
landing page
Updated Error message page with embedded link that opens in new tab too:
error page

File: dashboard.py [hunk 2]
```diff
@@ -83,7 +90,7 @@ def parse_contents(contents, filename):
         # will likely include categorical options in later instance (sooner)
     processed_df, cat_list = get_data(df, mapping)
     all_species = get_species_options(processed_df)
-    
+    print(mapping)
```
Comment (johnbradley) [inline dashboard.py:93]: Did you mean to leave this print in?

Comment (egrace479) [reply dashboard.py:93]: No, I thought I had removed all of those. Thanks for catching that.

**Review by johnbradley — COMMENTED** [review]
Is there a CSV I can test these changes with?

**Comment by egrace479** [comment]
> Is there a CSV I can test these changes with?

In `test_data`: 
- `HCGSD_base_filepath_noMap.csv`: has no lat/lon columns
- `HCGSD_base_filepath.csv`: has base filepath in `file_url` (there's also one with the full filepath, including filename)
- `Hoyal_Cuthill_GoldStandard_metadata_cleaned.csv`: doesn't have `file_url`

**Review by hlapp — APPROVED** [review]
LGTM! 👍

File: dashboard.py [hunk 3]
```diff
@@ -1,147 +1,149 @@
...
+                                    'font-size': '16px'}),
+                            id = 'upload-data',
+                            multiple = False
+                            ),
+                dcc.Store(id = 'memory'),
+                html.Hr(),
+                # Set up memory store, will revert on page refresh
+                #dcc.Store(id = 'memory'),
+                #html.Button('Process Data', 
+                #            id = 'memory-btn'),
```
Comment (johnbradley) [inline dashboard.py:31]: I'm wondering if this commented out code was intentionally left in.

File: dashboard.py [hunk 4]
```diff
@@ -1,147 +1,149 @@
...
+)
+
+def parse_contents(contents, filename):
+    '''
+    Function to read uploaded data.
+    '''
+    if contents is None:
+        raise PreventUpdate
+    content_type, content_string = contents.split(',')
+    #content_string = contents
```
Comment (johnbradley) [inline dashboard.py:59]: Another stray commented out line that might not be intensional.

**Review by johnbradley — APPROVED** [review]
LGTM. Just a couple commented out bits of code to consider removing.

File: components/query.py [hunk 5]
```diff
@@ -87,19 +93,20 @@ def get_images(df, subspecies, view, sex, hybrid, num_images):
     Imgs - List of html image elements with `src` element pointing to paths for the requested number of images matching given parameters.
            Returns html header4 "No Such Images. Please make another selection." if no images matching parameters exist.
     '''
-    filenames = get_filenames(df, subspecies, view, sex, hybrid, num_images)
+    filenames, filepaths = get_filenames(df, subspecies, view, sex, hybrid, num_images)
     if filenames == 0:
         #print("No Such Images. Please make another selection.")
```
Comment (thompsonmj) [inline components/query.py:98]: Could delete this one too.

**Review by thompsonmj — APPROVED** [review]

**Pull request merged**